libs/langchain-google-cloud-sql-pg/src/engine.ts

```diff
@@ -0,0 +1,106 @@
+import { AuthTypes, Connector, IpAddressTypes} from "@google-cloud/cloud-sql-connector";
+import { GoogleAuth } from "google-auth-library";
+import knex from "knex";
+import { getIAMPrincipalEmail } from "./utils/utils.js";
+
+class PostgresEngine {
+
+  private static _createKey = Symbol();
+  pool: knex.Knex<any, any[]>;
+  static connector: Connector;
+
+  constructor(key: Symbol , pool: knex.Knex<any, any[]>) {
+    if(key !== PostgresEngine._createKey) {
+      throw new Error("Only create class through 'create' method!")
+    }
+    this.pool = pool;
+  }
+
+  /**
+   * @param projectId GCP Project ID
+   * @param region Postgres Instance Region
+   * @param instance Postgres Instance name
+   * @param database Database name
+   * @param ipType IP address type. Defaults to IPAddressType.Public
+   * @param user Optional - Postgres user name. Defaults to undefined
+   * @param password Optional - Postgres user password. Defaults to undefined
+   * @param iamAccountEmail Optional - IAM service account email. Defaults to undefined
+   * @returns PostgressEngine instance
+   */
+
+  static async from_instance(
+      projectId: string,
+      region: string,
+      instance: string,
+      database: string,
+      ipType: IpAddressTypes = IpAddressTypes.PUBLIC,
+      user?: string, 
+      password?: string,
+      iamAccountEmail?: string,
+    ) {
+
+    let dbUser: string;
+    let enableIAMAuth: boolean;
+
+    if((!user && password) || (user && !password)) { // Trying to implement an XOR for strings
+      throw "Only one of 'user' or 'password' were specified. Either " +
+        "both should be specified to use basic user/password " +
+        "authentication or neither for IAM DB authentication.";
+    }
+
+    // User and password are given so we use the basic auth
+    if(user !== "" && password !== "") {
+      enableIAMAuth = false;
+      dbUser = user!;
+    } else {
+      enableIAMAuth = true;
+      if(iamAccountEmail !== undefined) {
+        dbUser = iamAccountEmail
+      } else {
+        // Get application default credentials
+        const auth = new GoogleAuth({
+          scopes: "https://www.googleapis.com/auth/cloud-platform"
+        });
+        // dbUser should be the iam principal email by passing the credentials obtained
+        dbUser = await getIAMPrincipalEmail(auth)
+      }
+    }
+
+    PostgresEngine.connector = new Connector();
```

Comment:
We will need to pass in a user agent for tracking. Can you look at the connector to see if this is possible. I will try to connect internally as well.

Reply:
I have been investigating but it seems that the node.js connector does not support the user agent as option parameter.